lib/livebook_web/live/user_component.ex

```diff
@@ -85,27 +88,38 @@ defmodule LivebookWeb.UserComponent do
 
   @impl true
   def handle_event("randomize_color", %{}, socket) do
-    data = %{
-      socket.assigns.data
-      | "hex_color" => User.random_hex_color(except: [socket.assigns.preview_user.hex_color])
-    }
-
-    handle_event("validate", %{"data" => data}, socket)
+    hex_color = User.random_hex_color(except: [socket.assigns.preview_user.hex_color])
+    handle_event("validate", %{"user" => %{"hex_color" => hex_color}}, socket)
   end
 
-  def handle_event("validate", %{"data" => data}, socket) do
-    {valid, user} =
-      case User.change(socket.assigns.user, data) do
-        {:ok, user} -> {true, user}
-        {:error, _errors, user} -> {false, user}
+  def handle_event("validate", %{"user" => attrs}, socket) do
+    params = Map.merge(socket.assigns.changeset.params, attrs)
+
+    changeset =
+      if socket.assigns.preview_user do
+        User.changeset(socket.assigns.preview_user, params)
+      else
+        User.changeset(socket.assigns.changeset, params)
       end
 
-    {:noreply, assign(socket, data: data, valid: valid, preview_user: user)}
+    {:noreply,
+     assign(socket, changeset: Map.put(changeset, :action, :validate), valid?: changeset.valid?)}
   end
 
-  def handle_event("save", %{"data" => data}, socket) do
-    {:ok, user} = User.change(socket.assigns.user, data)
-    Livebook.Users.broadcast_change(user)
-    {:noreply, socket}
+  def handle_event("save", %{"user" => params}, socket) do
+    changeset = User.changeset(socket.assigns.changeset, params)
```

Comment:
```suggestion
    changeset = Users.change_user(socket.assigns.changeset, params)
```

And `change_user` will set the action to validate.